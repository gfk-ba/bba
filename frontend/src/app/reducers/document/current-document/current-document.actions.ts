import { ApiDeleteRequest, ApiGetRequest, ApiPostRequest, ApiPutRequest } from '@dcs/ngx-tools';
import { Action } from '@ngrx/store';
import { currentDocumentManager } from './current-document.manager';
import { Document, IDocument } from '../models/document.class';

export class FetchCurrentDocument extends ApiGetRequest {
  constructor(id: string) {
    super(
      `documents/${id}`,
      currentDocumentManager.actions.fetch.base,
      currentDocumentManager.schema
    );
  }
}

export class CreateCurrentDocument extends ApiPostRequest {
  constructor(document: IDocument) {
    super(
      `documents`,
      currentDocumentManager.actions.create.base,
      document,
      currentDocumentManager.schema
    );
  }
}

export class UpdateCurrentDocument extends ApiPutRequest {
  constructor(document: Document) {
    super(
      `documents/${document.id}`,
      currentDocumentManager.actions.update.base,
      document.toObject(),
      currentDocumentManager.schema
    );
  }
}

export class DeleteCurrentDocument extends ApiDeleteRequest {
  constructor(document: Document) {
    super(
      `documents/${document.id}`,
      currentDocumentManager.actions.delete.base,
      document.toObject()
    );
  }
}

export class ResetCurrentDocument implements Action {
  readonly type = currentDocumentManager.actions.fetch.reset;
}
