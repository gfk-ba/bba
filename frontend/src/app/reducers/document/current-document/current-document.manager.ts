import { createSelector } from '@ngrx/store';
import { schema } from 'normalizr';
import { State } from '../../index';
import { currentProjectManager } from '../../project/current-project/current-project.manager';
import { projectsListStateSelector } from '../../project/projects-list/projects-list.manager';
import { Document, IDocument } from '../models/document.class';
import {
  crudEntityManagerFactory,
  INormalizedEntityState,
  generateNormalizedEntityState,
} from '@dcs/ngx-tools';

export interface ICurrentDocumentState extends INormalizedEntityState {
  entities: { documents: { [key: string]: IDocument } };
}

const initialState = generateNormalizedEntityState<ICurrentDocumentState>({ documents: {} });

const documentSchema = new schema.Entity(
  'documents',
  {
    project: currentProjectManager.schema,
    links: [new schema.Entity('documents')],
  },
  {
    idAttribute(entity) {
      return String(entity.id);
    },
    processStrategy(value: IDocument) {
      return {
        ...value,
        project: value.projectId,
        links: value.linkIds,
      };
    },
  }
);

const pureCurrentDocumentStateSelector = (state: State): ICurrentDocumentState => {
  return state.documents.currentDocument;
};

const currentDocumentStateSelector = createSelector(
  [pureCurrentDocumentStateSelector, projectsListStateSelector],
  (documentState, projectsState) => {
    return {
      ...documentState,
      entities: {
        documents: documentState.entities.documents,
        projects: projectsState.entities.projects,
      },
    };
  }
);

export const currentDocumentManager = crudEntityManagerFactory(
  'Current Document',
  initialState,
  currentDocumentStateSelector,
  documentSchema,
  Document
);
