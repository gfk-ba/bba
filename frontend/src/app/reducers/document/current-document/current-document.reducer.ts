import { Action } from '@ngrx/store';
import { currentDocumentManager, ICurrentDocumentState } from './current-document.manager';

export const currentDocument = (
  state: ICurrentDocumentState = currentDocumentManager.initialState,
  action: Action
): ICurrentDocumentState => {
  switch (
    action.type
    // overwrite or add any actions here, just default redux
  ) {
  }

  return currentDocumentManager.reducer(state, action);
};
