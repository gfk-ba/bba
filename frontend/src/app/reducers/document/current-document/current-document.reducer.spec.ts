import { currentDocument } from './current-document.reducer';

describe('currentDocumentReducer', () => {
  it('works', () => {
    expect(currentDocument(undefined, { type: 'any' })).toMatchSnapshot();
  });
});
