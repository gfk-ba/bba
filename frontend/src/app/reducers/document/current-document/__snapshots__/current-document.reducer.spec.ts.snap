// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`currentDocumentReducer works 1`] = `
Object {
  "entities": Object {
    "documents": Object {},
  },
  "error": null,
  "lastState": null,
  "loaded": false,
  "loading": false,
  "result": "",
  "updatedAt": null,
  "updating": false,
}
`;
